src/librustc_trans/trans/common.rs

```diff
@@ -534,6 +553,35 @@ impl<'a, 'tcx> FunctionContext<'a, 'tcx> {
                                          self.param_substs,
                                          value)
     }
+
+    /// If `type_needs_drop` returns false, then `ty` definitely has
+    /// no destructor attached. If it returns true, then `ty` is
+    /// definitely non-copy and might have a destructor attached,
+    pub fn type_needs_drop(&self, ty: Ty<'tcx>) -> bool {
```

Comment:
I think we should keep this method but refactor it so that it just calls to a top-level helper, so that it can be called from the drop glue as well (and hence the other, type-contents-based "type-may-drop" can go away in favor of this top-level helper). In other words I don't really like the idea of there being two functions that compute whether a drop is needed in slightly different ways. Seems like a recipe for confusion.